### PR TITLE
[FIRRTL] Parse and lower Open Aggregates Wires

### PR DESCRIPTION
**Pull request by seldridge**
Add support for parsing and lowering open aggregate wires. This currently works in a `firtool` pipeline so long as you run IMCP. I'm thinking about how to make it such that this optimization is not required for correctness (in `LowerClasses`). Currently `LowerClasses` will error out if the wires aren't removed.

There is fortunately already checking from `firrtl.propassign` that there are not multiple drivers.

Full example:

```
FIRRTL version 4.0.0
circuit Foo :
  module Foo :
    output a : { b : Integer, c: UInt<1>}[1]

    wire d : { e : Integer, f: UInt<1> }[1]
    invalidate d[0].f

    propassign d[0].e, Integer(1)
    propassign a[0].b, d[0].e

    connect d[0].f, UInt<1>(0)
    connect a[0].c, d[0].f
```

Produces with this patch (`firtool -ir-hw`):

```mlir
module {
  hw.module @Foo() -> (a_0_c: i1) {
    %false = hw.constant false
    hw.output %false : i1
  }
  om.class @Foo_Class() {
    %0 = om.constant 1 : si4
    om.class.field @a_0_b, %0 : si4
  }
}
```

### CODE REVIEW

**Review by dtzSiFive — COMMENTED** [review]
Cool! Happy to see LOA being built on!  Getting non-hw wire-like is something there's ask for, thanks for kicking this off!

Can we add and work on support for non-hw wires/storage as its own project separate from teaching LOA about them (which LGTM, thanks for the tests too!)?

(or if this is just the start, don't support parsing them until the pipeline behaves reasonably?)

I believe in "practice"/motivation is that this is meant to support wires that can and will be trivially dropped early (but this does not guarantee they will be), but worried this is opening a can of worms we're not quite ready for yet.

If wires can have non-hw, should nodes too?  If not, canonicalizers may need to be audited.

Passes may not expect non-hw in places that aren't ports, LowerXMR needs to handle this, so on? (IMCP/Inliner can be simplified, I think, woohoo!).  Dunno, but let's get our ducks in a row please :+1:.

**Comment by seldridge** [comment]
> I believe in "practice"/motivation is that this is meant to support wires that can and will be trivially dropped early (but this does not guarantee they will be), but worried this is opening a can of worms we're not quite ready for yet.

I can at least include (in this patch or in a follow-up) a canonicalizer, like `canonicalizeSingleSetConnect`, that will strip these when possible. I think it's reasonable to then make sure that passes properly error if they see non-HW types in wires.

Is there some more complicated work plan that you're envisioning?

Or: I'd rather open the can of worms now and see what breaks rather than trying to get things before enabling this. The one benefit of properties is that they are not load bearing, yet. Getting partial support in may be better than nothing.

> If wires can have non-hw, should nodes too? If not, canonicalizers may need to be audited.

Maybe it should be allowed? It would mean that the node "assignment" is equivalent to propassign and define of, respectively, properties and references. It's safest/easiest to disallow it. It's not required for the use case that I'm working with.

**Comment by dtzSiFive** [comment]
> I can at least include (in this patch or in a follow-up) a canonicalizer, like `canonicalizeSingleSetConnect`, that will strip these when possible. I think it's reasonable to then make sure that passes properly error if they see non-HW types in wires.

Okay, appreciate the defensive checks.

> Is there some more complicated work plan that you're envisioning?

Haha, alright man.  Nevermind.

> Or: I'd rather open the can of worms now and see what breaks rather than trying to get things before enabling this

Clearly.  GLHF.

**Review by dtzSiFive — APPROVED** [review]
Reviewed the LOA, added some nits/suggestions.

Still feels like introducing ability to have wire's throughout the pipeline that don't have HW in them should be introduced in its own change, but I see why it makes sense too to add LOA support while touching.

I'd forgotten we already support foreign types in wires ("why isn't this an IR change"), so happily that'll be better than I feared, I hope.  Still, non-base FIRRTLType's in storage may not be well-handled in the pipeline.  We'll see!  Probes probably get the worst of it, but at least ""yet"" probes in wires isn't load-bearing or expected to actually work in non-basic scenarios.

For scenarios where we can't remove the wires, the conversion to OM may need a little work but hopefully is okay re:graph regions.

File: lib/Dialect/FIRRTL/Transforms/LowerOpenAggs.cpp [hunk 1]
```diff
@@ -581,6 +582,48 @@ LogicalResult Visitor::visitDecl(InstanceOp op) {
   return success();
 }
 
+LogicalResult Visitor::visitDecl(WireOp op) {
+  auto pmi =
+      mapPortType(op.getResultTypes()[0], op.getLoc(), op.getInnerSymAttr());
+  if (failed(pmi))
+    return failure();
+  PortMappingInfo portMapping = *pmi;
```
[the remark this answers is not in the document]
Comment (seldridge) [reply lib/Dialect/FIRRTL/Transforms/LowerOpenAggs.cpp:590]: Will add this. Are you alright if I rework the `port-mapping-info` print into something more YAML-ish?

I'll also rename this to something that doesn't mention ports in an NFC commit.

File: lib/Dialect/FIRRTL/Transforms/LowerOpenAggs.cpp [hunk 2]
```diff
@@ -581,6 +582,48 @@ LogicalResult Visitor::visitDecl(InstanceOp op) {
...
+            .getResult();
+  }
+
+  // Create the non-HW wires.
+  for (auto &[type, fieldID, _, suffix] : portMapping.fields)
+    nonHWValues[FieldRef(op.getResult(), fieldID)] =
+        builder
+            .create<WireOp>(
+                type, builder.getStringAttr(Twine(op.getName()) + suffix),
+                op.getNameKind(), builder.getArrayAttr({}), StringAttr{}, false)
```
[the remark this answers is not in the document]
Comment (seldridge) [reply lib/Dialect/FIRRTL/Transforms/LowerOpenAggs.cpp:616]: Yes, can be elided. Will update this.

File: lib/Dialect/FIRRTL/Transforms/LowerOpenAggs.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/Transforms/LowerOpenAggs.cpp:590]: Awesome, and yes please! Would love it to get the @seldridge debug treatment!

And totally, renaming sounds good and your plan sounds great.  Thanks! 🙏

[the remark this answers is not in the document]
Comment (seldridge) [reply lib/Dialect/FIRRTL/Transforms/LowerOpenAggs.cpp:590]: New output added. E.g., here is a new format:

```
===- Running Lower Open Aggregates Pass -------------------------------------===
Visiting modules:
  - name: "Bar"
    ports:
      - name: "x"
        type: !firrtl.openvector<openbundle<y: integer, z: uint<1>>, 1>
        mapping:
          hardware: !firrtl.vector<bundle<z: uint<1>>, 1>
          non-hardware:
            - type: !firrtl.integer
              fieldID: 2
              isFlip: false
              suffix: "_0_y"
          mappedToNull:
          newSym: <empty>
    body:
  - name: "Foo"
    ports:
      - name: "a"
        type: !firrtl.openvector<openbundle<b: integer, c: uint<1>>, 1>
        mapping:
          hardware: !firrtl.vector<bundle<c: uint<1>>, 1>
          non-hardware:
            - type: !firrtl.integer
              fieldID: 2
              isFlip: false
              suffix: "_0_b"
          mappedToNull:
          newSym: <empty>
    body:
      - wire:
          name: "d"
          type: !firrtl.openvector<openbundle<e: integer, f: uint<1>>, 1>
            mapping:
              hardware: !firrtl.vector<bundle<f: uint<1>>, 1>
              non-hardware:
                - type: !firrtl.integer
                  fieldID: 2
                  isFlip: false
                  suffix: "_0_e"
              mappedToNull:
              newSym: <empty>
      - instance:
          name: "bar"
          module: @Bar
          ports:
            - name: "x"
              type: !firrtl.openvector<openbundle<y: integer, z: uint<1>>, 1>
              mapping:
                hardware: !firrtl.vector<bundle<z: uint<1>>, 1>
                non-hardware:
                  - type: !firrtl.integer
                    fieldID: 2
                    isFlip: false
                    suffix: "_0_y"
                mappedToNull:
                newSym: <empty>
```

File: test/Dialect/FIRRTL/lower-open-aggs.mlir [hunk 3]
```diff
@@ -233,3 +233,38 @@ firrtl.circuit "BundleOfProps" {
...
+    %b_c = firrtl.opensubfield %b[c] : !firrtl.openbundle<c: string>
+    firrtl.propassign %b_c, %0 : !firrtl.string
+
+    // CHECK-NEXT: %d_0 = firrtl.wire : !firrtl.string
+    // CHECK-NEXT: firrtl.propassign %d_0, %b_c
+    %d = firrtl.wire : !firrtl.openvector<string, 1>
+    %d_0 = firrtl.opensubindex %d[0] : !firrtl.openvector<string, 1>
+    firrtl.propassign %d_0, %b_c : !firrtl.string
+  }
+}
```
[the remark this answers is not in the document]
Comment (seldridge) [reply test/Dialect/FIRRTL/lower-open-aggs.mlir:258]: Will add this in a `fixup!` commit / force push.

File: test/Dialect/FIRRTL/lower-open-aggs.mlir [hunk 4]
```diff
@@ -233,3 +233,38 @@ firrtl.circuit "BundleOfProps" {
...
+// -----
+
+// CHECK-LABEL: circuit "WireSymbols"
+firrtl.circuit "WireSymbols" {
+  // CHECK-LABEL: module{{.*}} @WireSymbols
+  firrtl.module @WireSymbols() {
+    // CHECk-NEXT: %a = firrtl.wire sym [<@sym_a_c, 1, public>] : !firrtl.bundle<c: uint<1>>
+    // CHECk-NEXT: %a_b = firrtl.wire : !firrtl.string
+    %a = firrtl.wire sym [<@sym_a_c, 2, public>] : !firrtl.openbundle<b: string, c: uint<1>>
+  }
```
[the remark this answers is not in the document]
Comment (seldridge) [reply test/Dialect/FIRRTL/lower-open-aggs.mlir:269]: Will add this in a force push. Yeah, this won't really do anything different given that there are no connect-like operations on aggregates which have flips in them. However, the test is still good to have!

File: lib/Dialect/FIRRTL/Transforms/LowerOpenAggs.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/Transforms/LowerOpenAggs.cpp]: Wow!! Amazing, thank you! :grin:

File: test/Dialect/FIRRTL/lower-open-aggs.mlir (hunk 4 above)
[the remark this answers is not in the document]
Comment (dtzSiFive) [reply test/Dialect/FIRRTL/lower-open-aggs.mlir:310]: Thanks, appreciate the coverage :+1: .

**Pull request merged**